src/defrag.c

```diff
@@ -1103,18 +1127,37 @@ void activeDefragCycle(void) {
             }
 
             if (!defrag_later_item_in_progress) {
-                /* Scan the keyspace dict unless we're scanning the expire dict. */
-                if (!expires_cursor)
-                    cursor = kvstoreDictScanDefrag(db->keys, slot, cursor,
-                                                   defragScanCallback,
+                if (defrag_state == DEFRAG_KEYS) {
+                    cursor = kvstoreDictScanDefrag(db->keys, slot, cursor, defragScanCallback,
                                                    &defragfns, &ctx);
-                /* When done scanning the keyspace dict, we scan the expire dict. */
-                if (!cursor)
+                    if (!cursor) defrag_state = DEFRAG_EXPIRES;
+                }
+
+                if (defrag_state == DEFRAG_EXPIRES) {
                     expires_cursor = kvstoreDictScanDefrag(db->expires, slot, expires_cursor,
-                                                           scanCallbackCountScanned,
-                                                           &defragfns, NULL);
+                                                           scanCallbackCountScanned, &defragfns, NULL);
+                    if (!expires_cursor) defrag_state = DEFRAG_KEYS;
```

Comment:
when done with expires, aren't we suppose to move to pubsub?

Reply:
yeah, thanks.
i'm still using python to test it, i'll gifure out a way to test it in the unit tests that it can fit in at lease large-memory.

Reply:
done.